MyApplication/app/src/main/java/com/example/myapplication/ui/main/view/UserActivity.kt

```diff
@@ -0,0 +1,82 @@
+package com.example.myapplication.ui.main.view
+
+import android.os.Bundle
+import android.view.View
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProviders
+import androidx.recyclerview.widget.DividerItemDecoration
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.myapplication.R
+import com.example.myapplication.data.api.ApiHelper
+import com.example.myapplication.data.api.RetrofitBuilder
+import com.example.myapplication.data.model.User
+import com.example.myapplication.ui.base.ViewModelFactory
+import com.example.myapplication.ui.main.adapter.MainAdapter
+import com.example.myapplication.ui.main.viewmodel.MainViewModel
+import com.example.myapplication.utils.Status
+import kotlinx.android.synthetic.main.activity_user.*
+
+class UserActivity : AppCompatActivity() {
+
+    private lateinit var viewModel: MainViewModel
+    private lateinit var adapter: MainAdapter
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_user)
+        setupViewModel()
+        setupUI()
+        setupObservers()
+    }
+
+    private fun setupViewModel() {
+        viewModel = ViewModelProviders.of(
+            this,
+            ViewModelFactory(ApiHelper(RetrofitBuilder.apiService))
+        ).get(MainViewModel::class.java)
+    }
+
+    private fun setupUI() {
+        recyclerView.layoutManager = LinearLayoutManager(this)
+        adapter = MainAdapter(arrayListOf())
+        recyclerView.addItemDecoration(
+            DividerItemDecoration(
+                recyclerView.context,
+                (recyclerView.layoutManager as LinearLayoutManager).orientation
+            )
+        )
+        recyclerView.adapter = adapter
+    }
+
+    private fun setupObservers() {
+        viewModel.getUsers().observe(this, Observer {
+            it?.let { resource ->
+                when (resource.status) {
+                    Status.SUCCESS -> {
+                        recyclerView.visibility = View.VISIBLE
+                        progressBar.visibility = View.GONE
+                        resource.data?.let { users -> retrieveList(users) }
+                    }
+                    Status.ERROR -> {
+                        recyclerView.visibility = View.VISIBLE
+                        progressBar.visibility = View.GONE
+                        Toast.makeText(this, it.message, Toast.LENGTH_LONG).show()
+                    }
+                    Status.LOADING -> {
+                        progressBar.visibility = View.VISIBLE
+                        recyclerView.visibility = View.GONE
+                    }
+                }
```

Comment:
This is business logic of your app. It handles different states. You should be able to test it. That's why you have to move it somewhere in that way to be able to cover it in unit tests

Reply:
I see no problem in using this logic here as I already call getUsers() function from ViewModel where business logic actually is.
The example I followed: https://stackoverflow.com/questions/57966523/where-to-put-business-logic-when-working-with-mvvm